src/v/cluster/members_backend.cc

```diff
@@ -593,10 +598,30 @@ void members_backend::calculate_reallocations_batch(
                 continue;
             }
         }
+        size_t rand_idx = 0;
+        /**
+         * We use 64 bit field initialized with random number to sample topic
+         * partitions to move, every time the bitfield is exhausted we
+         * reinitialize it and reset counter. This way we chose random
+         * partitions to move without iterating multiple times over the topic
+         * set
+         */
+        std::bitset<sizeof(uint64_t)> sampling_bytes;
         for (const auto& p : metadata.get_assignments()) {
             if (idx_it->second > current_idx++) {
                 continue;
             }
+
+            if (rand_idx % sizeof(uint64_t) == 0) {
+                sampling_bytes = random_generators::get_int(
+                  std::numeric_limits<uint64_t>::max());
+                rand_idx = 0;
+            }
+
+            if (sampling_bytes.test(rand_idx++)) {
+                continue;
+            }
+
```

Comment:
q: Should we not `id_it->second++` regardless of whether the sampling picks this assignment so that we always start from where we left off? Move L625 before L615

Otherwise we can potentially rescan the assignments because of holes with sampling.

Reply:
good point